lib/Optimizer/Transforms/GenKernelExecution.cpp

```diff
@@ -1392,6 +1393,91 @@ class GenerateKernelExecution
+    auto ptrPtrTy = cudaq::cc::PointerType::get(i8PtrTy);
+    auto cast1 = builder.create<cudaq::cc::CastOp>(loc, ptrPtrTy, buffer);
+    auto ptr3Ty = cudaq::cc::PointerType::get(ptrPtrTy);
+    auto stdVec0 = builder.create<cudaq::cc::CastOp>(loc, ptr3Ty, stdVec);
+    builder.create<cudaq::cc::StoreOp>(loc, cast1, stdVec0);
+    auto cast2 = builder.create<cudaq::cc::CastOp>(loc, i64Ty, buffer);
+    auto endBuff = builder.create<arith::AddIOp>(loc, cast2, buffSize);
+    auto cast3 = builder.create<cudaq::cc::CastOp>(loc, ptrPtrTy, endBuff);
+    auto stdVec1 = builder.create<cudaq::cc::ComputePtrOp>(
+        loc, ptr3Ty, stdVec, ArrayRef<cudaq::cc::ComputePtrArg>{1});
+    builder.create<cudaq::cc::StoreOp>(loc, cast3, stdVec1);
+    auto stdVec2 = builder.create<cudaq::cc::ComputePtrOp>(
+        loc, ptr3Ty, stdVec, ArrayRef<cudaq::cc::ComputePtrArg>{2});
+    builder.create<cudaq::cc::StoreOp>(loc, cast3, stdVec2);
+    auto zero = builder.create<arith::ConstantIntOp>(loc, 0, 64);
+    auto nullPtr = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, zero);
+
+    // 2) Iterate over the arguments passed in and populate the vector.
+    SmallVector<BlockArgument> blockArgs{dropAnyHiddenArguments(
+        hostFuncEntryBlock->getArguments(), devFuncTy, addThisPtr)};
+    for (auto iter : llvm::enumerate(blockArgs)) {
+      std::int32_t i = iter.index();
+      auto pos = builder.create<cudaq::cc::ComputePtrOp>(
+          loc, ptrPtrTy, buffer, ArrayRef<cudaq::cc::ComputePtrArg>{i});
+      auto blkArg = iter.value();
+      if (isa<cudaq::cc::PointerType>(blkArg.getType())) {
+        auto castArg = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, blkArg);
+        builder.create<cudaq::cc::StoreOp>(loc, castArg, pos);
+        continue;
+      }
+      auto temp = builder.create<cudaq::cc::AllocaOp>(loc, blkArg.getType());
+      builder.create<cudaq::cc::StoreOp>(loc, blkArg, temp);
+      auto castTemp = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, temp);
+      builder.create<cudaq::cc::StoreOp>(loc, castTemp, pos);
+    }
+
+    auto resultBuffer = builder.create<cudaq::cc::AllocaOp>(loc, i8PtrTy);
+    builder.create<cudaq::cc::StoreOp>(loc, nullPtr, resultBuffer);
+    auto castResultBuffer =
+        builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, resultBuffer);
+    auto castStdvec = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, stdVec);
+    Value loadKernName = builder.create<LLVM::AddressOfOp>(
+        loc, cudaq::opt::factory::getPointerType(kernelNameObj.getType()),
+        kernelNameObj.getSymName());
+    auto castKernelNameObj =
+        builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, loadKernName);
+    builder.create<func::CallOp>(
+        loc, std::nullopt, cudaq::runtime::launchKernelVersion2FuncName,
+        ArrayRef<Value>{castKernelNameObj, castStdvec, castResultBuffer});
+
+    // FIXME: Drop any results on the floor for now and return random data left
```

Comment:
I thought we support returning values from kernels in c++ and python now?

Reply:
I'll add this to the new flow in a later PR.

Reply:
As to the question, it's nuanced.

Does the kernel launch have the _capability_ to handle return values? Yes. Is that capability fully functional and correct? Almost completely, if you consider C++. Do any of the kernel launchers deal with return values? Almost never. There is a singular path that can return a `int`, `double` and `bool` (only when remote simulation and other qualifiers IIRC).